subiquity/controllers/filesystem.py

```diff
@@ -131,71 +128,50 @@ def add_disk_partition(self, disk):
+        log.debug('model has bootable device? {}'.format(system_bootable))
+        if not system_bootable and len(disk._partitions) == 0:
+            if self.is_uefi():
+                log.debug('Adding EFI partition first')
+                part = self.model.add_partition(disk=disk, partnum=1, size=UEFI_GRUB_SIZE_BYTES, flag='bios_grub')
+                fs = self.model.add_filesystem(part, 'fat32')
+                self.model.add_mount(fs, '/boot/efi')
             else:
-                current_disk.add_partition(partnum=spec["partnum"],
-                                           size=spec["bytes"],
-                                           fstype=spec["fstype"],
-                                           mountpoint=spec["mountpoint"])
-        except Exception:
-            log.exception('Failed to add disk partition')
-            log.debug('Returning to add-disk-partition')
-            # FIXME: on failure, we should repopulate input values
-            self.add_disk_partition(disk)
+                log.debug('Adding grub_bios gpt partition first')
+                part = self.model.add_partition(disk=disk, partnum=1, size=BIOS_GRUB_SIZE_BYTES, flag='bios_grub')
+            disk.grub_device = True
+
+            # adjust downward the partition size to accommodate
+            # the offset and bios/grub partition
+            # XXX should probably only do this if the partition is now too big to fit on the disk?
+            log.debug("Adjusting request down:" +
+                      "{} - {} = {}".format(spec['bytes'], part.size,
+                                            spec['bytes'] - part.size))
+            spec['bytes'] -= part.size
+            spec['partnum'] = 2
+
+        part = self.model.add_partition(disk=disk, partnum=spec["partnum"], size=spec["bytes"])
+        if spec['fstype'] is not None:
+            fs = self.model.add_filesystem(part, spec['fstype'])
+            if spec['mountpoint']:
+                self.model.add_mount(fs, spec['mountpoint'])
 
         log.info("Successfully added partition")
-
-        log.debug("FS Table: {}".format(current_disk.get_fs_table()))
         self.prev_view()
 
-    def add_disk_format_handler(self, disk, spec):
-        log.debug('add_disk_format_handler')
-        current_disk = self.model.get_disk(disk)
-        log.debug('format spec: {}'.format(spec))
-        log.debug('disk.freespace: {}'.format(current_disk.freespace))
-        current_disk.format_device(spec['fstype'], spec['mountpoint'])
-        log.debug("FS Table: {}".format(current_disk.get_fs_table()))
+    def add_format_handler(self, volume, spec):
+        log.debug('add_format_handler')
+        if spec['fstype'] is not None:
+            fs = self.model.add_filesystem(volume, spec['fstype'])
+        else:
+            fs = volume._fs
+        if spec['mountpoint']:
+            if fs is None:
+                raise Exception("%s is not formatted" % (volume.path,))
```

Comment:
Eh, I don't know. This error case is a logic bug that's going to crash the installer so I don't know if it's worth the effort of coming up with a custom exception hierarchy for that...

Changed to .format() though.